### PR TITLE
Always use cancelSubscription form for cancelling recurring contributions

### PR DESCRIPTION
**Pull request by mattwire**
Overview
----------------------------------------
Replaces #16717, follow up to #17146.

Includes #22784 and #21895.

There was existing code to handle cancellation if the processor does not support cancellation.

You may still want to record cancellation reason, see instructions to manually cancel with your provider and send a notification email to the contributor.

But if the processor reports that it does not `supportCancelRecurring` the form is never displayed and the built in entity enable/disable via javascript is used. This only really makes sense if there is no payment processor.

Before
----------------------------------------
enable/disable javascript used if processor does not support cancelling recur (cancel link still available in list of recurs!).

After
----------------------------------------
Always display the cancellation form. Eg. for a recur with payment processor:
image

And without a payment processor:
image

Technical Details
----------------------------------------

Comments
----------------------------------------
This cleans up another inconsistency in the way civi handles things. @eileenmcnaughton This finishes off the cancellation stuff we were working on a few versions back.

### CODE REVIEW

**Comment by eileenmcnaughton** [comment]
I'm just going through triaging for 'needs test' & ok without test. I guess the logic is PRs need a test unless we can't see how they could be tested or we know there is existing relevant cover. I think this could be tested.

**Comment by eileenmcnaughton** [comment]
I took a look at this since a new PR got raised that addresses some of the same code / makes at least one change that is the same & put up a PR to address the 2 things that stalled this PR https://github.com/civicrm/civicrm-core/pull/18790

**Comment by seamuslee001** [comment]
Test fail relates i believe

**Comment by eileenmcnaughton** [comment]
Yep - there is another open PR related to these links which is where discussion is taking place  #18787  - the tldr is that we also need to think about the situation where the appropriate links are different for someone dealing with their own recurring vs someone else's

**Comment by mattwire** [comment]
test this please

**Comment by mattwire** [comment]
test this please

**Comment by mattwire** [comment]
test this please

**Comment by mattwire** [comment]
```
CRM_Contribute_Page_TabTest::testLinks
Failed asserting that two strings are equal.
--- Expected
+++ Actual
@@ @@
-'<span><a href="/index.php?q=civicrm/contact/view/contributionrecur&amp;reset=1&amp;id=1&amp;cid=3&amp;context=contribution" class="action-item crm-hover-button" title='View Recurring Payment' >View</a><a href="/index.php?q=civicrm/contribute/updaterecur&amp;reset=1&amp;action=update&amp;crid=1&amp;cid=3&amp;context=contribution" class="action-item crm-hover-button" title='Edit Recurring Payment' >Edit</a><a href="/index.php?q=civicrm/contribute/unsubscribe&amp;reset=1&amp;crid=1&amp;cid=3&amp;context=contribution" class="action-item crm-hover-button" title='Cancel' >Cancel</a></span>'
+'<span><a href="/index.php?q=civicrm/contact/view/contributionrecur&amp;reset=1&amp;id=1&amp;cid=3&amp;context=contribution" class="action-item crm-hover-button" title='View Recurring Payment' >View</a><a href="/index.php?q=civicrm/contribute/unsubscribe&amp;reset=1&amp;crid=1&amp;cid=3&amp;context=contribution" class="action-item crm-hover-button" title='Cancel' >Cancel</a></span>'

/home/jenkins/bknix-dfl/build/core-18196-7nonk/web/sites/all/modules/civicrm/tests/phpunit/CRM/Contribute/Page/TabTest.php:55
/home/jenkins/bknix-dfl/build/core-18196-7nonk/web/sites/all/modules/civicrm/tests/phpunit/CiviTest/CiviUnitTestCase.php:226
/home/jenkins/bknix-dfl/extern/phpunit7/phpunit7.phar:615
```

**Comment by mattwire** [comment]
test this please

**Comment by mattwire** [comment]
We can build on this test: https://github.com/civicrm/civicrm-core/pull/21588

File: CRM/Contribute/Page/Tab.php [hunk 1]
```diff
@@ -72,16 +72,13 @@ public static function recurLinks(int $recurID, $context = 'contribution') {
...
     if ($paymentProcessorObj->supports('cancelRecurring')) {
-      unset($links[CRM_Core_Action::DISABLE]['extra'], $links[CRM_Core_Action::DISABLE]['ref']);
-      $links[CRM_Core_Action::DISABLE]['url'] = "civicrm/contribute/unsubscribe";
-      $links[CRM_Core_Action::DISABLE]['qs'] = "reset=1&crid=%%crid%%&cid=%%cid%%&context={$context}";
+      $links[CRM_Core_Action::DISABLE] = [
+        'name' => ts('Cancel'),
+        'title' => ts('Cancel'),
+        'url' => 'civicrm/contribute/unsubscribe',
+        'qs' => "reset=1&crid=%%crid%%&cid=%%cid%%&context={$context}",
+      ];
```
Comment (jitendrapurohit) [inline CRM/Contribute/Page/Tab.php:81]: Does the above change mean we load `Cancel` link only for processors supporting `cancelRecurring`?

But from the description -

>After
>Always display the cancellation form. Eg. for a recur with payment processor:

it seems the cancel link should always be visible and condition should be checked on the form, i.e, for `Send cancellation request to processor` field? Is that correct?

With this PR applied, I see the Cancel link is not displayed for processors not having `supportsCancelRecurring()`.

image

@mattwire can you pls clarify?

Comment (mattwire) [reply CRM/Contribute/Page/Tab.php]: @jitendrapurohit You are right. I've have pushed a fix.

**Comment by BettyDolfing** [comment]
Test this please

**Comment by jaapjansma** [comment]
@mattwire can you explain what this PR changes? We see the exact same behaviour in https://dmaster.demo.civicrm.org (which does not have this patch). See screenshot below:

Screenshot_20220411_161552

**Comment by eileenmcnaughton** [comment]
@mattwire I've always shied away from looking at this because of the unexplained changes in `getAllProcessors` - but the part where you alter the action makes sense to me - so I pulled that part out into a separate PR that I think you can probably merge 
https://github.com/civicrm/civicrm-core/pull/23210

**Comment by eileenmcnaughton** [comment]
@mattwire pr template needs updating now - to explain what the point is of this change / why the PR still exists

**Comment by mattwire** [comment]
@eileenmcnaughton Yes, not sure that this bit is still required.

**Pull request closed without merging**